test/client_reset_test.dart

```diff
@@ -0,0 +1,256 @@
+        ).clientResetHandler.clientResyncMode,
+        ClientResyncModeInternal.discardLocal);
+    expect(
+        Configuration.flexibleSync(
+          user,
+          [Task.schema, Schedule.schema],
+          clientResetHandler: const RecoverUnsyncedChangesHandler(),
+        ).clientResetHandler.clientResyncMode,
+        ClientResyncModeInternal.recover);
+
+    expect(
+        Configuration.flexibleSync(
+          user,
+          [Task.schema, Schedule.schema],
+          clientResetHandler: const RecoverOrDiscardUnsyncedChangesHandler(),
+        ).clientResetHandler.clientResyncMode,
+        ClientResyncModeInternal.recoverOrDiscard);
+
+    expect(Configuration.flexibleSync(user, [Task.schema, Schedule.schema]).clientResetHandler.clientResyncMode, ClientResyncModeInternal.recoverOrDiscard);
+  });
+
+  baasTest('ManualRecoveryHandler error is reported in callback', (appConfig) async {
+    final app = App(appConfig);
+    final user = await getIntegrationUser(app);
+
+    final resetCompleter = Completer<ClientResetError>();
+    final config = Configuration.flexibleSync(
+      user,
+      [Task.schema, Schedule.schema],
+      clientResetHandler: ManualRecoveryHandler((syncError) {
+        resetCompleter.complete(syncError);
+      }),
+    );
+
+    final realm = await Realm.open(config);
+    await realm.syncSession.waitForUpload();
+
+    await triggerClientReset(realm);
+
+    final error = await resetCompleter.future;
+    expect(error.message, contains('Bad client file identifier'));
+  });
+
+  for (Type clientResetHandlerType in [
+    RecoverOrDiscardUnsyncedChangesHandler,
+    RecoverUnsyncedChangesHandler,
+    DiscardUnsyncedChangesHandler,
+  ]) {
+    baasTest('$clientResetHandlerType.manualResetFallback invoked when throw an error on Before Callback', (appConfig) async {
+      final app = App(appConfig);
+      final user = await getIntegrationUser(app);
+
+      final onManualResetFallback = Completer<void>();
+      final config = Configuration.flexibleSync(user, [Task.schema, Schedule.schema],
+          clientResetHandler: Creator.create(
+            clientResetHandlerType,
+            beforeResetCallback: (beforeFrozen) => throw Exception("This fails!"),
+            manualResetFallback: (clientResetError) => onManualResetFallback.complete(),
+          ));
+
+      final realm = await Realm.open(config);
+      await realm.syncSession.waitForUpload();
+
+      await triggerClientReset(realm);
+
+      await onManualResetFallback.future;
+    });
+  }
+
+  for (Type clientResetHandlerType in [
+    RecoverOrDiscardUnsyncedChangesHandler,
+    RecoverUnsyncedChangesHandler,
+    DiscardUnsyncedChangesHandler,
+  ]) {
+    baasTest('$clientResetHandlerType.manualResetFallback invoked when throw an error on After Callbacks', (appConfig) async {
+      final app = App(appConfig);
+      final user = await getIntegrationUser(app);
+
+      final onManualResetFallback = Completer<void>();
```

Comment:
Similar to the previous test, let's assert on the exception we get in the manual reset fallback.